### ISSUE TITLE
Build fails with libjpeg-turbo 3.0.3: Undefined symbols for architecture x86_64: "_jpeg_std_message_table"

### ISSUE THREAD

**Issue by barracuda156** [Beep6581/RawTherapee#7074]
Looks that `libjpeg-turgo` changed API in 3.0.3, and this code now does not work:
https://github.com/Beep6581/RawTherapee/blob/5a25619bd50d42059abe9818420f639b6d10e3bf/rtengine/jdatasrc.cc#L251-L262

Windows build has this issue years ago: https://github.com/Beep6581/RawTherapee/issues/4713
Now identical linking failure on macOS: https://build.macports.org/builders/ports-14_x86_64-builder/builds/39199/steps/install-port/logs/stdio

This compiles though:
```
const char * const jpeg_std_message_table[] = {
#include "jerror.h"
  NULL
};
```
(Functionality not tested.)

**Comment by barracuda156** [comment]
Builds fine against 3.0.2, confirmed. So definitely `libjpeg-turbo` issue.

**Comment by ryandesign** [comment]
libjpeg-turbo says it is a RawTherapee issue; see https://github.com/libjpeg-turbo/libjpeg-turbo/issues/763

**Comment by dcommander** [comment]
libjpeg-turbo did ***not*** change its API.  `jpeg_std_message_table[]` was never a part of any API, either public or private.  It is a local symbol in **jerror.c**, and its exposure at the global level was only accidental and only on certain platforms.  Our Un*x shared library uses a linker script to determine symbol exposure, so I posit that RawTherapee's approach could not have worked with that library.  (In other words, your approach was not cross-platform.)

tl;dr: Your software relies on accidental, undocumented, unofficial behavior that only occurred on certain platforms, so you can't complain about that behavior changing.  There is almost certainly a better way to accomplish what you're trying to accomplish than to needlessly duplicate a lot of the internals of libjpeg-turbo.

**Comment by dcommander** [comment]
Note that, when you call `jpeg_std_error()`, a pointer to `jpeg_std_message_table[]` is returned in `err->jpeg_message_table`.  That is the correct way to access `jpeg_std_message_table[]`.  After calling `jpeg_std_error()`, your code can then override the `error_exit()`, `emit_message()`, `output_message()`, `format_message()`, or `reset_error_mgr()` methods if necessary.

**Comment by barracuda156** [comment]
@dcommander Just to be clear, this is not _my_ software, we did not add this part in MacPorts, we just got a sudden breakage upon `libjpeg-turbo` update.

**Comment by dcommander** [comment]
> @dcommander Just to be clear, this is not _my_ software, we did not add this part in MacPorts, we just got a sudden breakage upon `libjpeg-turbo` update.

Yes, you've already mentioned that.  But this issue is under the RawTherapee project, so "your" refers to the RawTherapee authors.

**Comment by Benitoite** [comment]
Currently RT builds on Mac ARM but only with libjpeg-turbo downgraded to 3.0.2 
https://github.com/libjpeg-turbo/libjpeg-turbo/issues/763#issuecomment-2106286608 says RT was/is using an unexposed call of some sort.

**Comment by kmilos** [comment]
Note that MSYS2 libjpeg-turbo just got bumped to 3.0.3.

**Comment by Benitoite** [comment]
I found a patch written for this bug: https://github.com/termux-user-repository/tur/pull/1027

**Comment by Lawrence37** [comment]
Thank you @Benitoite! I will have a look at your pull request.

**Comment by haplo** [comment]
Will there be a minor release (5.10.1) that includes a fix for this bug, or will it wait until 5.11?

I ask because RawTherapee is currently broken on Arch Linux, and depending on the expected time for a fixed release I would consider opening an issue with them to include this patch temporarily.

Thank you for your support!

**Comment by haplo** [comment]
Nevermind, Arch Linux just released a new version with the jpeg-turbo patch:

https://gitlab.archlinux.org/archlinux/packaging/packages/rawtherapee/-/commit/ca57665a9b247386c78a46585ca977370396133f

**Comment by BenLand100** [comment]
I have the same question as @haplo --- this issue is preventing rawtherapee from starting or building for downstream users, requiring maintainers to patch. One would think this would prompt a timely new release...?

**Comment by Benitoite** [comment]
@BenLand100 The next release will be in a few hours/days.  cf. https://github.com/Beep6581/RawTherapee/issues/6966#issuecomment-2297937410

**Comment by Lawrence37** [comment]
If we had more contributors, we might consider a release. The amount of time involved in publishing a new release is too much considering that a simple `git cherry-pick -n efdc5bc 7789a85` patches the code.
